### PR TITLE
[Cherry-pick][v0.18.0.post1][Perf] Optimize Wan2.2 device free on image preprocess

### PR DESCRIPTION
**Pull request by fan2956**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Cherry-pick https://github.com/vllm-project/vllm-omni/pull/2852

This PR fixes two critical issues in pipeline_wan2_2_i2v.py to prevent CPU fallback and device mismatches:
1. Moves input tensor to the target device before video_processor.preprocess to eliminate CPU-side preprocessing.
2. Optimizes mask_lat_size generation by using PyTorch native tensor slicing instead of list indexing, ensuring full execution on the target device (NPU/GPU) and avoiding device free.
No functional changes are made to the original logic. The changes improve inference stability across hardware platforms.

## Test Plan
```
export MULTI_STREAM_MEMORY_REUSE=2
vllm serve /home/y00958577/Wan2.2-I2V-A14B-Diffusers/ \
  --omni \
  --port 8099 \
  --usp 8 \
  --use-hsdp \
  --enforce-eager \
  --log-stats \
  --profiler-config '{"profiler": "torch", "torch_profiler_dir": "./vllm_profile", "torch_profiler_with_stack": "False"}'\
  --vae-patch-parallel-size 8 \
  --vae-use-tiling
```

```
curl -X POST http://localhost:8099/v1/videos \
    -F "prompt='一只棕色野兔的正面特写镜头，采用低角度仰拍视角，营造亲密而庄严的视觉冲击。兔子一双圆润漆黑的大眼睛直视镜头深处，眼神中交织着野生动物的警觉与一>丝难以言喻的温柔好奇，仿佛在与观者建立跨越物种的静默对话。它毛色呈现层次丰富的棕褐渐变，从浅奶油色腹部过渡到深棕背部，每根毛发纹理清晰可辨，在侧光下泛着丝绸般>的光泽。细长洁白的胡须共有三对，随呼吸节奏微微颤动，偶尔因捕捉气流信息而轻轻摇摆。
两只标志性的长耳完全竖立，耳廓外侧覆盖短密棕毛，内侧则露出粉嫩的血管网络，薄如蝉翼的皮肤下血液流动隐约可见，耳朵以细微幅度不时转动，精准定位声源方向。背景是一>片澄澈的蔚蓝天空，形态蓬松的白色积云以缓慢速度横向漂移，云影在兔子头顶交替变化，光线随之明暗流转。晴朗天气的明媚阳光从画面左上方45度角倾泻而下，在兔脸右侧形成>柔和的伦勃朗式阴影，强化了面部立体感和皮毛质感。
兔子湿润的黑鼻子持续进行每秒三至四次的快速抽动，这是它们感知化学信号的本能动作，粉色三瓣嘴随之轻启，露出正在反刍的洁白门齿，下颌以稳定节奏左右研磨。摄影采用大>光圈浅景深，焦点牢牢锁定在兔子双眼连线所在的焦平面，背景天空和远景绿色植被虚化成圆润的彩色光斑，前景几根嫩绿草叶闯入画面边缘，以缓慢弧线随风摇曳，暗示着和煦的>春日微风。
整个场景弥漫着宁静致远的田园诗意，色彩温暖饱和，充满生命力。兔子在持续五秒的对视后，以典型 lagomorph 特征完成一次完整的瞬膜眨眼——第三眼睑从内侧横向滑过眼球，继
而缓缓歪头向右十五度，这个行为在动物行为学中代表认知加工和好奇心表达，耳朵随之向同一方向倾斜，最终恢复正视姿态，胡须舒展，完成这段短暂而珍贵的自然纪录。'" \
    -F "input_reference=@/home/zf/vllm-omni/rabbit.jpeg" \
    -F "size=832x480" \
    -F "seconds=5" \
    -F "fps=16" \
    -F "num_frames=81" \
    -F "guidance_scale=1" \
    -F "guidance_scale_2=1" \
    -F "flow_shift=5.0" \
    -F "num_inference_steps=4" \
    -F "seed=42"
```

## Test Result
before:
e2e_total_ms = 9,837.386 
after:
e2e_total_ms = 9,512.703

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [√] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [√] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [√] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by lishunyang12 — CHANGES_REQUESTED** [review]
## Review: [Perf] Optimize Wan2.2 device free on image preprocess v018

### Summary

The mask tensor optimization (Part 2) is clean and correct. Creating `mask_lat_size` directly on the target device and using native slicing (`1:` instead of `list(range(1, num_frames))`) is a good improvement that avoids an unnecessary CPU-to-device transfer.

However, the image preprocessing changes (Part 1) have a bug that needs to be fixed before merging.

---

### Bug: Wrong variable in `last_image` block (copy-paste error)

```python
if last_image is not None:
    if isinstance(last_image, PIL.Image.Image):
        image = TF.to_tensor(image).to(device)  # <-- BUG: should be `last_image`, not `image`
        last_image_tensor = video_processor.preprocess(last_image, height=height, width=width)
```

Two problems here:
1. **Wrong variable converted**: `image` is converted to a tensor instead of `last_image`. At this point `image` has already been processed earlier, so this re-converts it for no reason.
2. **Converted result is unused**: The tensor is assigned back to `image`, but `last_image` (still a PIL Image) is what gets passed to `video_processor.preprocess`. So this line has no effect on `last_image_tensor` and the stated goal of moving preprocessing to the device is not achieved for `last_image`.

This should be:
```python
last_image = TF.to_tensor(last_image).to(device)
last_image_tensor = video_processor.preprocess(last_image, height=height, width=width)
```

### Minor concern: Semantic change in `preprocess` behavior

By converting `image` from `PIL.Image.Image` to a `torch.Tensor` before calling `video_processor.preprocess`, the code now takes the tensor branch inside `preprocess` instead of the PIL branch. While both branches ultimately produce a normalized tensor, the PIL path includes explicit `convert_to_rgb` and specific PIL-based resize logic, whereas the tensor path skips RGB conversion and uses `torch.nn.functional.interpolate` for resize. This is likely fine in practice (since `TF.to_tensor` preserves channels), but worth being aware of -- it is a subtle behavioral change, not a pure no-op optimization.

### Part 2 (mask optimization): Approved

The changes to `prepare_latents` are correct and well-motivated:
- `torch.ones(..., device=latent_condition.device)` avoids creating on CPU then transferring.
- `mask_lat_size[:, :, 1:]` is cleaner and faster than `mask_lat_size[:, :, list(range(1, num_frames))]`.
- Removing the trailing `.to(latent_condition.device)` is correct since the tensor is already on the right device.

### Recommendation

Please fix the `last_image` copy-paste bug, then this is good to merge.

**Comment by hsliuustc0106** [comment]
BLOCKER scan:
- Correctness: ISSUES (line 472: copy-paste error - uses `image` instead of `last_image`)
- Reliability/Safety: PASS
- Breaking Changes: PASS
- Test Coverage: PASS (e2e latency provided)
- Documentation: PASS
- Security: PASS

OVERALL: 1 BLOCKER FOUND

VERDICT: REQUEST_CHANGES

There is a copy-paste error on line 472:

```python
if isinstance(last_image, PIL.Image.Image):
    image = TF.to_tensor(image).to(device)  # BUG: should be `last_image`, not `image`
    last_image_tensor = video_processor.preprocess(last_image, ...)
```

This means `last_image` is never converted to a tensor or moved to the target device.

Please fix line 472 to:
```python
last_image = TF.to_tensor(last_image).to(device)
```

Good improvement removing the redundant `.to(latent_condition.device)` on the last line since device is already set on line 795.

File: vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2_i2v.py [hunk 1]
```diff
@@ -459,6 +460,7 @@ def forward(
         video_processor = VideoProcessor(vae_scale_factor=self.vae_scale_factor_spatial)
 
         if isinstance(image, PIL.Image.Image):
+            image = TF.to_tensor(image).to(device)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2_i2v.py:463]: This changes the input to `video_processor.preprocess` from a PIL Image to a Tensor, taking a different code path that uses different resize logic. Where is the validation that this preserves generation quality? Run benchmarks to ensure no accuracy degradation.

File: vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2_i2v.py [hunk 2]
```diff
@@ -467,6 +469,7 @@ def forward(
         # Handle last_image if provided
         if last_image is not None:
             if isinstance(last_image, PIL.Image.Image):
+                image = TF.to_tensor(last_image).to(device)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2_i2v.py:472]: This still assigns the tensor to `image` instead of `last_image`. As a result, `last_image_tensor` is still processed from the original PIL image on the CPU, and the earlier `image` variable is incorrectly overwritten.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
The `last_image` bugfix still assigns to the wrong variable. Additionally, since this is a performance optimization PR that changes the preprocessing branch from PIL to tensor, please provide evidence that generation quality is preserved. Finally, where are the memory measurements? Please provide before/after memory measurements to validate the device free optimization claims.

**Review by gcanlin — APPROVED** [review]
